### PR TITLE
Ensure resource destruction order in ReadFromStorageStep.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix the order of destruction for resources in `ReadFromStorage` step of query plan. It might cause crashes in rare cases. Possibly connected with #15610.

### CODE REVIEW

File: src/Processors/Pipe.h [hunk 1]
```diff
@@ -101,7 +101,7 @@ class Pipe
     void setQuota(const std::shared_ptr<const EnabledQuota> & quota);
 
     /// Do not allow to change the table while the processors of pipe are alive.
-    void addTableLock(const TableLockHolder & lock) { holder.table_locks.push_back(lock); }
+    void addTableLock(TableLockHolder lock) { holder.table_locks.emplace_back(std::move(lock)); }
```
Comment (Akazz) [inline src/Processors/Pipe.h:104]: If return value of `emplace_back(&&)` is not used at the call site it is equivalent to `push_back(&&)`.

File: src/Processors/QueryPlan/ReadFromStorageStep.cpp [hunk 2]
```diff
@@ -13,33 +13,26 @@ namespace DB
...
+    size_t max_block_size,
+    size_t max_streams)
 {
     /// Note: we read from storage in constructor of step because we don't know real header before reading.
     /// It will be fixed when storage return QueryPlanStep itself.
 
+    /// Move arguments into stack in order to ensure order of destruction in case of exception.
+    auto context = std::move(context_);
+    auto storage = std::move(storage_);
+    auto table_lock = std::move(table_lock_);
```
Comment (Akazz) [inline src/Processors/QueryPlan/ReadFromStorageStep.cpp:34]: I suggest we do not do this, but instead move around c-tor's parameters accordingly. And so we would be totally safe

File: src/Processors/Pipe.h (hunk 1 above)
Comment (KochetovNicolai) [reply src/Processors/Pipe.h:104]: Sure

File: src/Processors/QueryPlan/ReadFromStorageStep.cpp (hunk 2 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/ReadFromStorageStep.cpp:34]: But destruction order of parameters is not specified.

**Review by Akazz — APPROVED** [review]
lgtm

File: src/Processors/QueryPlan/ReadFromStorageStep.cpp (hunk 2 above)
Comment (Akazz) [reply src/Processors/QueryPlan/ReadFromStorageStep.cpp]: That is true. But the actual release of resources will occur somewhere in the calling code.

**Comment by KochetovNicolai** [comment]
fuzzer is reported here: #15683

**Pull request merged**